src/listwidget.cpp

```diff
@@ -191,16 +191,24 @@ void ListWidget::update_scroll_offset(std::uint32_t pos)
 			} else { // "pos" is towards the beginning of the list; don't scroll
 				set_scroll_offset(0);
 			}
+			return;
 		}
 
 		// Check if items at the top of the "context" are visible. If not,
 		// we'll have to scroll up.
 		if (pos < cur_scroll_offset + num_context_lines) {
 			if (pos >= num_context_lines) {
-				set_scroll_offset(pos - num_context_lines);
+				const std::uint32_t target_offset = pos - num_context_lines;
+				set_scroll_offset(std::min(target_offset, max_offset));
```

Comment:
> So the bug was caused by us calling set_scroll_offset without checking if its argument exceeds max_offset. 

No, at first I thought so too.
However, when I was debugging this issue, it looked like none of the `set_scroll_offset` calls were hit.
This happened with the following values in the old implementation of `update_scroll_offset`:

| var | value |
|---|---|
| h | 7 |
| cur_scroll_offset | 32 |
| num_lines | 33 |
| max_offset | 26 |
| pos | 32 |

I mainly added the extra `std::min(_, max_offset)` calls to guarantee that after returning, we will never be scrolled too far down.

Reply:
> I suggest moving the check into set_scroll_offset then, so it can't be missed when we add more calls. This also has the benefit of simplifying the code of this method a little bit, as we no longer need max_offset, calls to std::min, and the new safeguard if at the end of this block. What do you think?

Just realized this idea is still valid.
I already merged, so I won't change it right now (but might include this suggested change in a future PR)